### PR TITLE
Blocks and token matching

### LINKED ISSUE #2159: ci( yi( di( etc does not work if parenthesis contains a string literal with a "("
Trying to use vi( or any _i( combination inside a parenthesis containing a "(" inside a string literal does not work, for example trying to copy the statement in the if () does not work:

`if ("hello".IndexOf('(') == 0)`

### PR DESCRIPTION
**Pull request by ricksladkey**
### Changes

- Convert `i(` (paren block motion), etc. and `%` (token matching) to use the same infrastructure
- Fix paren blocks, etc. to handle string and character literals
- Add tests for block and token matching corner cases

### Fixes

- Fixes #2159

### Discussion

There is a long comment in `GetBlock` discussing how vim behaves and how VsVim handles the problem. The idea is not to behave exactly as vim does, because it is not hard to trip up vim either, but to handle the most common cases that previously tripped up VsVim.

### CODE REVIEW

File: Src/VimCore/MotionUtil.fs [hunk 1]
```diff
@@ -348,6 +348,210 @@ type MatchingTokenKind =
...
+        // Is the char at the given point escaped?
+        let isEscaped point =
+            match SnapshotPointUtil.TrySubtractOne point with
+            | None -> false
+            | Some point -> SnapshotPointUtil.GetChar point = '\\'
+
+        // Is the char at the given point double escaped?
+        let isDoubleEscaped point =
+            match SnapshotPointUtil.TrySubtract point 2, SnapshotPointUtil.TrySubtract point 2 with
+            | None, _
```
Comment (jaredpar) [inline Src/VimCore/MotionUtil.fs]: Huh, didn't realize F# had this type of fall through behavior.

Comment (ricksladkey) [reply Src/VimCore/MotionUtil.fs]: The underscore matches a tuple. In any case, I didn't change that code; I just had to move it to a new type for reuse. So you probably wrote the original code!
